src/Corax/IndexWriter.cs

```diff
@@ -677,25 +649,144 @@ public void Commit()
+                var entries = _buffer[fieldId][term];
+                if (fieldTree.TryGetValue(termsSpan, out var existing, out var encodedKey) == false)
+                {
+                    if(AddNewTerm(entries,  tmpBuf, out  termId))
+                        fieldTree.Add(termsSpan, termId, encodedKey);
+                    continue;
+                }
+
+                if(AddEntriesToTerm(tmpBuf, existing, llt, entries, out  termId))
+                    fieldTree.Add(termsSpan, termId, encodedKey);
+
+            }
+        }
+
+        private bool AddEntriesToTerm(Span<byte> tmpBuf, long existing, LowLevelTransaction llt, List<long> entries, out long termId)
+        {
+            if ((existing & (long)TermIdMask.Set) != 0)
+            {
+                var id = existing & ~0b11;
+                var setSpace = Container.GetMutable(llt, id);
+                ref var setState = ref MemoryMarshal.AsRef<SetState>(setSpace);
+                var set = new Set(llt, Slices.Empty, setState);
+                entries.Sort();
+                set.Add(entries);
+                setState = set.State;
+                termId = -1;
+                return false; // it's already there, no need
+            }
+            else
+            {
+                if ((existing & (long)TermIdMask.Small) != 0)
+                {
+                    var id = existing & ~0b11;
+                    var smallSet = Container.Get(llt, id).ToSpan();
+                    // combine with existing value
+                    var cur = 0L;
+                    ZigZagEncoding.Decode<int>(smallSet, out var pos); // discard the first entry, the count
+                    while (pos < smallSet.Length)
+                    {
+                        var value = ZigZagEncoding.Decode<long>(smallSet, out var len, pos);
+                        pos += len;
+                        cur += value;
+                        entries.Add(cur);
+                    }
+
+                    Container.Delete(llt, _postingListContainerId, id);
+                    return AddNewTerm(entries, tmpBuf, out  termId);
+                }
+                // single
+                
+                if (entries.Count == 1 && entries[0] == existing)
+                {
+                    // Same element to add, nothing to do here.
+                    termId = -1;
+                    return false;
+                }
+
+                entries.Add(existing);
+                return AddNewTerm(entries, tmpBuf, out  termId);
+            }
+        }
+
+        private unsafe void InsertNumericFieldLongs(Tree fieldsTree, int fieldId, Span<byte> tmpBuf)
+        {
+            FixedSizeTree fieldTree = fieldsTree.FixedTreeFor(_fieldsMapping.GetByFieldId(fieldId).FieldNameLong, sizeof(long));
+            var llt = Transaction.LowLevelTransaction;
+            var sortedTerms = _bufferLongs[fieldId].Keys.ToArray();// TODO: let's avoid this allocation
```

Comment:
We haven't done allocations optimizations for the most part. Dont worry about it.